Adds a --flatnames option to mlabconfig.py

nkinkade opened the pull request:
Currently, the ndt_ssl blackbox_exporter checks are failing. One reason is that the target hostnames are not using the "flattened" domain which is compatible with our wildcard SSL certificate for measurement-lab.org.  This PR adds a `--flatnames` options to mlabconfig.py which can instruct Prom ta0rget generator functions to return flattened domain names suitable for TLS/SSL wildcard certs.

One small change in `generate_prometheus_targets.sh` takes advantage of this new functionality to fix the targets for the `tcp_v4_tls_online` module for  the NDT SSL check.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/m-lab/operator/171)
<!-- Reviewable:end -->

stephen-soltesz commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> -- good catch.

---

Reviewed 3 of 3 files at r1.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[plsync/mlabconfig.py, line 233 at r1](https://reviewable.io:443/reviews/m-lab/operator/171#-Kz9fpykjXQkZTnDq3rb:-Kz9fpykjXQkZTnDq3rc:bl3bul6) ([raw file](https://github.com/m-lab/operator/blob/116164938c93eb7732f81d32db3af61fb5131ebf/plsync/mlabconfig.py#L233)):*
> ```Python
>         '',
>         '--flatnames',
>         dest='flatnames',
> ```

The flag name is okay, but as noted below, I suggest using `use_flatnames` as the dest variable.

---

*[plsync/mlabconfig.py, line 655 at r1](https://reviewable.io:443/reviews/m-lab/operator/171#-Kz9ep6W07Z9pZSVF-kN:-Kz9ep6W07Z9pZSVF-kO:bdc3ho7) ([raw file](https://github.com/m-lab/operator/blob/116164938c93eb7732f81d32db3af61fb5131ebf/plsync/mlabconfig.py#L655)):*
> ```Python
> def select_prometheus_experiment_targets(experiments, select_regex,
>                                          target_templates, common_labels,
>                                          rsync_only, flatnames):
> ```

a readability suggestion: Maybe call this `use_flatnames`? that name is connotes a boolean value more strongly than `flatnames`. In particular, I might equally expect `flatnames` to be a list of strings.

---

*[plsync/mlabconfig.py, line 686 at r1](https://reviewable.io:443/reviews/m-lab/operator/171#-Kz9fhQoGq9W87JLYiIO:-Kz9fhQoGq9W87JLYiIP:b-pyh356) ([raw file](https://github.com/m-lab/operator/blob/116164938c93eb7732f81d32db3af61fb5131ebf/plsync/mlabconfig.py#L686)):*
> ```Python
>             host = experiment.hostname(node)
>             # Don't use the flatten_hostname() function in this module because
>             # it adds too much overhead. Just replace the first three dots with
> ```

What does overhead mean in this context?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/operator/171#-:-Kz9g7K-XEHDXkZ3Y79M:bhu5p39)*
<!-- Sent from Reviewable.io -->

nkinkade commented:
PTAL?

---

Review status: 1 of 3 files reviewed at latest revision, 3 unresolved discussions.

---

*[plsync/mlabconfig.py, line 233 at r1](https://reviewable.io:443/reviews/m-lab/operator/171#-Kz9fpykjXQkZTnDq3rb:-Kz9q_HjOQgxXcMPiJR_:butox17) ([raw file](https://github.com/m-lab/operator/blob/116164938c93eb7732f81d32db3af61fb5131ebf/plsync/mlabconfig.py#L233)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

The flag name is okay, but as noted below, I suggest using `use_flatnames` as the dest variable.
</blockquote></details>

I've change dest variable, but I also changed the option name too since I think it's confusing to have a destination variable be different than the option, when it can be avoided.

---

*[plsync/mlabconfig.py, line 655 at r1](https://reviewable.io:443/reviews/m-lab/operator/171#-Kz9ep6W07Z9pZSVF-kN:-Kz9qlCsyIM-XnEKXwZ6:bt2vr1b) ([raw file](https://github.com/m-lab/operator/blob/116164938c93eb7732f81d32db3af61fb5131ebf/plsync/mlabconfig.py#L655)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

a readability suggestion: Maybe call this `use_flatnames`? that name is connotes a boolean value more strongly than `flatnames`. In particular, I might equally expect `flatnames` to be a list of strings.
</blockquote></details>

Fixed.

---

*[plsync/mlabconfig.py, line 686 at r1](https://reviewable.io:443/reviews/m-lab/operator/171#-Kz9fhQoGq9W87JLYiIO:-Kz9qmB8ILjz8hvEFLsH:bc4yebg) ([raw file](https://github.com/m-lab/operator/blob/116164938c93eb7732f81d32db3af61fb5131ebf/plsync/mlabconfig.py#L686)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

What does overhead mean in this context?
</blockquote></details>

I literally spent 20 to 30 minutes last night thinking about how to accomplish this making use of the function you already wrote named `flatten_hostnames()`.  It's not hard to do, and there are many ways someone could accomplish it. However, I couldn't come up with an options that didn't require 4 or 5 lines of code, or really long operations that become unreadable like `host.split()[:-2].join[5]()-do_this.function[1]`. I went back and forth on usages of `split`, `rsplit`, `replace`, regexes, etc. And then finally I said: why go through all this when I can accomplish this very concisely in a single line? It does hardcode the domain (measurement-lab.org) in the code, but I think this is a safe assumption for years to come.

Do you agree, or would you prefer that I fall back on the existing function?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/operator/171#-:-Kz9rtpDqEJCpZsQEfm3:b-833305)*
<!-- Sent from Reviewable.io -->

stephen-soltesz commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[plsync/mlabconfig.py, line 686 at r1](https://reviewable.io:443/reviews/m-lab/operator/171#-Kz9fhQoGq9W87JLYiIO:-Kz9u_xpDg4oygbSlw2O:b8kxspf) ([raw file](https://github.com/m-lab/operator/blob/116164938c93eb7732f81d32db3af61fb5131ebf/plsync/mlabconfig.py#L686)):*
<details><summary><i>Previously, nkinkade wrote…</i></summary><blockquote>

I literally spent 20 to 30 minutes last night thinking about how to accomplish this making use of the function you already wrote named `flatten_hostnames()`.  It's not hard to do, and there are many ways someone could accomplish it. However, I couldn't come up with an options that didn't require 4 or 5 lines of code, or really long operations that become unreadable like `host.split()[:-2].join[5]()-do_this.function[1]`. I went back and forth on usages of `split`, `rsplit`, `replace`, regexes, etc. And then finally I said: why go through all this when I can accomplish this very concisely in a single line? It does hardcode the domain (measurement-lab.org) in the code, but I think this is a safe assumption for years to come.

Do you agree, or would you prefer that I fall back on the existing function?
</blockquote></details>

Simpler is almost always better. I think you made the right choice. "overhead" implies something different in my head. "code complexity" is the phrase I would use to describe what you were wrestling with here. We're all good here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/operator/171#-:-Kz9uw51Rnilm2Sp9Fys:bnfp4nl)*
<!-- Sent from Reviewable.io -->
